src/infra/dex/okx/mod.rs

```diff
@@ -0,0 +1,255 @@
+        // Safe to unwrap as HMAC can take key of any size
+        let mut mac = HmacSha256::new_from_slice(self.api_secret_key.as_bytes()).unwrap();
+        mac.update(data.as_bytes());
+        let signature = mac.finalize().into_bytes();
+
+        BASE64_STANDARD.encode(signature)
+    }
+
+    pub async fn swap(
+        &self,
+        order: &dex::Order,
+        slippage: &dex::Slippage,
+    ) -> Result<dex::Swap, Error> {
+        let query = self.defaults.clone().with_domain(order, slippage);
+        let quote = {
+            // Set up a tracing span to make debugging of API requests easier.
+            // Historically, debugging API requests to external DEXs was a bit
+            // of a headache.
+            static ID: AtomicU64 = AtomicU64::new(0);
+            let id = ID.fetch_add(1, atomic::Ordering::Relaxed);
+            self.quote(&query)
+                .instrument(tracing::trace_span!("quote", id = %id))
+                .await?
+        };
+
+        let quote_result = quote.data.first().ok_or(Error::NotFound)?;
+
+        let (input, output, max_sell_amount) = match order.side {
+            order::Side::Buy => (
+                eth::Asset {
+                    token: order.buy,
+                    amount: quote_result.router_result.from_token_amount,
+                },
+                eth::Asset {
+                    token: order.sell,
+                    amount: quote_result.router_result.to_token_amount,
+                },
+                slippage.add(quote_result.router_result.from_token_amount),
+            ),
+            order::Side::Sell => (
+                eth::Asset {
+                    token: order.sell,
+                    amount: quote_result.router_result.from_token_amount,
+                },
+                eth::Asset {
+                    token: order.buy,
+                    amount: quote_result.router_result.to_token_amount,
+                },
+                quote_result.router_result.from_token_amount,
+            ),
+        };
+
+        Ok(dex::Swap {
+            call: dex::Call {
+                to: eth::ContractAddress(quote_result.tx.to),
+                calldata: quote_result.tx.data.clone(),
+            },
+            input,
+            output,
+            allowance: dex::Allowance {
+                spender: eth::ContractAddress(quote_result.tx.from),
+                amount: dex::Amount::new(max_sell_amount),
+            },
+            gas: eth::Gas(quote_result.tx.gas), // todo ms: increase by 50% according to docs?
+        })
+    }
+
+    async fn quote(&self, query: &dto::SwapRequest) -> Result<dto::SwapResponse, Error> {
+        let request_builder = self
+            .client
+            .request(reqwest::Method::GET, self.endpoint.clone())
+            .query(query);
+
+        let quote = util::http::roundtrip!(
+            <dto::SwapResponse, dto::Error>;
+            request_builder,
+            |request| {
```

Comment:
Is it not possible to add these headers to the builder before putting it into the `roundtrip` macro to avoid introducing extra complexity?